### PR TITLE
vcmi: fix test compiling and passing

### PR DESCRIPTION
**Pull request by rilian-la-te**
It is merge of #1623 and #2067

Notable differences:

1. Scripting (at least Lua) tests passing (and correctly enabled only if lua enabled)
2. All tests passing with H3 data

### CODE REVIEW

**Comment by rilian-la-te** [comment]
@IvanSavenko @Nordsoft91

**Review by IvanSavenko — COMMENTED** [review]
Have doubts as to how useful tests are in their current state (even if h3 data issue gets resolved). 

Might change opinion if they'll actually catch a bug at some point.

File: lib/mapping/CMapService.h [hunk 1]
```diff
@@ -24,72 +24,75 @@ class IMapPatcher;
 class ModCompatibilityInfo;
 
 /**
- * The map service provides loading and saving of VCMI/H3 map files.
+ * The map service provides loading of VCMI/H3 map files. It can
+ * be extended to save maps later as well.
  */
-class DLL_LINKAGE CMapService
+class DLL_LINKAGE IMapService
```
[the remark this answers is not in the document]
Comment (rilian-la-te) [reply lib/mapping/CMapService.h:30]: This is copied from #2067, maybe it is partial revert, I do not know.

**Comment by rilian-la-te** [comment]
@IvanSavenko tests already caught some bugs (some 1.2 PRs was made by tests outcome, for example). But yes, tests need to be largely expanded to cover all mechanics (not small part as they currently cover).

**Comment by rilian-la-te** [comment]
@IvanSavenko Anyway, found abandoned mine serialization bug now with this PR) So, will fix)

**Comment by rilian-la-te** [comment]
@IvanSavenko @Nordsoft91 all tests passing with H3 data now.

**Comment by rilian-la-te** [comment]
> Have doubts as to how useful tests are in their current state (even if h3 data issue gets resolved).

Even now I found one bug with abandoned mine loading (introduced by Hota PR). Fix is included here.

**Review by Nordsoft91 — COMMENTED** [review]
I'll download branch and test it locally

**Comment by rilian-la-te** [comment]
@Nordsoft91 I tested it on ArchLinux on Clang compiler, if it should be relevant, via ```ctest --preset linux-clang-test```

**Comment by Nordsoft91** [comment]
What does it mean?
```
-- Performing Test CONV
-- Performing Test CONV - Success
-- Performing Test COPY
-- Performing Test COPY - Failed
-- Performing Test MOVE_ASSIGN
-- Performing Test MOVE_ASSIGN - Failed
-- Configuring done
-- Generating done
```

**Comment by rilian-la-te** [comment]
@Nordsoft91 we need to check some compiler flags (warnings mostly). Google test is not -Wall compilant, so, we check compiler warnings, and if check is success, then we add -Wno<that warning>. For now there is 3 warnings checked:

```
-Wimplicit-int-float-conversion
-Wdeprecated-copy-with-user-provided-copy
-Wno-virtual-move-assign
```

First two is clang-only, third is GCC-only.

**Comment by rilian-la-te** [comment]
relevant CMake code:

```
check_cxx_compiler_flag(-Wimplicit-int-float-conversion CONV)
if(CONV)
	add_compile_options(-Wno-error=implicit-int-float-conversion) #Used in googletest
endif()
check_cxx_compiler_flag(-Wdeprecated-copy-with-user-provided-copy COPY)
if(COPY)
	add_compile_options(-Wno-deprecated-copy-with-user-provided-copy) #Used in googletest
endif()
check_cxx_compiler_flag(-Wvirtual-move-assign MOVE_ASSIGN)
if(MOVE_ASSIGN)
	add_compile_options(-Wno-virtual-move-assign) #GCC is too strict here
endif()
```

**Comment by Nordsoft91** [comment]
Still 2 asserts hit:
- CModHandler.cpp, 103
- ResourceID.cpp, 207

**Comment by rilian-la-te** [comment]
@Nordsoft91 I fixed modhandler assert, and I cannot reproduce second assert.

**Review by Nordsoft91 — APPROVED** [review]
Then ok, I'll not checkout branch again, but if another assert will be reproduced we can take a look later

**Pull request merged**